src/rules/no-element-overwrite.ts

```diff
@@ -0,0 +1,162 @@
+  isLiteral,
+  isIdentifier,
+  isCallExpression,
+} from "../utils/nodes";
+
+const rule: Rule.RuleModule = {
+  create(context: Rule.RuleContext) {
+    return {
+      SwitchCase(node: Node) {
+        const switchCase = node as SwitchCase;
+        checkStatements(switchCase.consequent);
+      },
+
+      BlockStatement(node: Node) {
+        const block = node as BlockStatement;
+        checkStatements(block.body);
+      },
+
+      Program(node: Node) {
+        const program = node as Program;
+        checkStatements(program.body);
+      },
+    };
+
+    function checkStatements(statements: Array<Statement | ModuleDeclaration>) {
+      const usedKeys: Map<string, KeyWriteCollectionUsage> = new Map();
+      let collection: Node | null = null;
+      statements.forEach(statement => {
+        const keyWriteUsage = getKeyWriteUsage(statement);
+        if (keyWriteUsage) {
+          if (collection && !areEquivalent(keyWriteUsage.collectionNode, collection, context.getSourceCode())) {
+            usedKeys.clear();
+          }
+          const sameKeyWriteUsage = usedKeys.get(keyWriteUsage.indexOrKey);
+          if (sameKeyWriteUsage) {
+            context.report({
+              node: keyWriteUsage.node,
+              message: 'Verify this is the index that was intended; "{{index}}" was already set on line {{line}}.',
+              data: {
+                index: keyWriteUsage.indexOrKey,
+                line: String(sameKeyWriteUsage.node.loc!.start.line),
+              },
+            });
+          }
+          usedKeys.set(keyWriteUsage.indexOrKey, keyWriteUsage);
+          collection = keyWriteUsage.collectionNode;
+        } else {
+          usedKeys.clear();
+        }
+      });
+    }
+
+    function getKeyWriteUsage(node: Node): KeyWriteCollectionUsage | undefined {
+      if (isExpressionStatement(node)) {
+        const expr = node.expression;
+        return arrayKeyWriteUsage(expr) || mapOrSetKeyWriteUsage(expr);
+      }
+    }
+
+    function arrayKeyWriteUsage(node: Node): KeyWriteCollectionUsage | undefined {
+      // a[b] = ...
+      if (isSimpleAssignment(node) && isMemberExpression(node.left) && node.left.computed) {
+        const lhs = node.left as MemberExpression;
+        const index = extractIndex(lhs.property);
+        if (!index || isUsed(lhs.object, node.right)) return;
+
+        return {
+          collectionNode: lhs.object,
+          indexOrKey: index,
+          node: lhs.object,
+        };
+      }
+    }
+
+    function mapOrSetKeyWriteUsage(node: Node): KeyWriteCollectionUsage | undefined {
+      if (isCallExpression(node) && isMemberExpression(node.callee)) {
+        const propertyAccess = node.callee;
+        if (isIdentifier(propertyAccess.property)) {
+          const methodName = propertyAccess.property.name;
+          const addMethod = methodName === "add" && node.arguments.length == 1;
+          const setMethod = methodName === "set" && node.arguments.length == 2;
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: == should be === ([typescript:S1440](https://sonarcloud.io/organizations/sonarsource/rules#rule_key=typescript%3AS1440))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=eslint-plugin-sonarjs&pullRequest=36&issues=AWK0wBF4jr7wspRoK9BC&open=AWK0wBF4jr7wspRoK9BC)

[ISSUE_KEY]: AWK0wBF4jr7wspRoK9BC